### ISSUE TITLE
System.Linq.Enumerable was broken in .NET 5 rc 1 wasm runtime

### ISSUE THREAD

**Issue by ElderJames** [dotnet/runtime#42536]
### Describe the bug

When I update from Blazor 3.2 to .Net 5 RC 1 and then System.Linq.Enumerable was broken in wasm runtime.

It works well in debugging.

### To Reproduce

Code: https://github.com/ant-design-blazor/ant-design-blazor/tree/master/components/table
Online: https://ant-design-blazor.github.io/en-US/components/table

### Exceptions (if any)

![image](https://user-images.githubusercontent.com/7550366/93735260-f5b14a80-fc0e-11ea-8801-64feaba40375.png)

### Further technical details
- ASP.NET Core version .NET 5 rc 1
- Include the output of `dotnet --info`
- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version

**Comment by mkArtakMSFT** [comment]
@lewing  FYI

**Comment by lewing** [comment]
@marek-safar this looks suspiciously like the linker

**Comment by lewing** [comment]
@ElderJames can you verify that disabling linking by adding 

```
<PropertyGroup>
  <BlazorWebAssemblyEnableLinking>false</BlazorWebAssemblyEnableLinking>
</PropertyGroup>
```
To your project works around the issue?

**Comment by ElderJames** [comment]
Thanks @lewing , 

I tried, but it was still the exception.

**Comment by marek-safar** [comment]
This is libraries bug by not annotating/respecting reflection call correctly. I think it goes down to https://github.com/dotnet/runtime/blob/c48127daec460692ee18f7ae7552bcc3f7fa8499/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L218 which is not recognized by linker and issues warning which is ignored by libraries build.

This is a regression from 3.2 release

/cc @eerhardt

**Comment by pranavkm** [comment]
For a 5.0 app, you can turn off trimming by configuring `<PublishTrimmed>false</PublishTrimmed>` in your project file.

**Comment by eerhardt** [comment]
@ElderJames - do you have a branch of the repo that is targeting `net5.0`?

**Comment by lewing** [comment]
> Thanks @lewing ,
> 
> I tried, but it was still the exception.

Whoops, I gave the wrong instuctions, try https://github.com/dotnet/runtime/issues/42536#issuecomment-696802055 instead

**Comment by eerhardt** [comment]
I can repro this error with this simple change to the `FetchData` page of a `dotnet new blazorwasm` template app:

```diff
    protected override async Task OnInitializedAsync()
    {
        forecasts = await Http.GetFromJsonAsync<WeatherForecast[]>("sample-data/weather.json");

+        forecasts = forecasts.AsQueryable().Take(2).ToArray();
    }
```

```
blazor.webassembly.js:1 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Arg_NullReferenceException
System.NullReferenceException: Arg_NullReferenceException
   at System.Linq.EnumerableRewriter.FindEnumerableMethod(String name, ReadOnlyCollection`1 args, Type[] typeArgs)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.EnumerableQuery`1[[BlazorApp3.Pages.FetchData.WeatherForecast, BlazorApp3, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].GetEnumerator()
   at System.Linq.EnumerableQuery`1[[BlazorApp3.Pages.FetchData.WeatherForecast, BlazorApp3, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Collections.Generic.LargeArrayBuilder`1[[BlazorApp3.Pages.FetchData.WeatherForecast, BlazorApp3, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[WeatherForecast](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[WeatherForecast](IEnumerable`1 source)
   at BlazorApp3.Pages.FetchData.OnInitializedAsync()
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)
```

Working on a fix.

**Comment by eerhardt** [comment]
The issue here is that when you use methods on `System.Linq.Queryable` against an in-memory collection, they get re-written to the corresponding method on `System.Linq.Enumerable`. However, the linker doesn't know about this connection between the two types, since the connection is made using Reflection here:

https://github.com/dotnet/runtime/blob/9c472545df220d7ae26329e1649f880c4ee127e8/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L39-L45

There's 2 options I can see that will fix this.

1. We can annotate the above `FindEnumerableMethod` method with a `DynamicallyAccessedMembers` attribute, telling it to preserve methods on `System.Linq.Enumerable`.
    - The issue there is ALL the methods on `System.Linq.Enumerable` will be preserved as soon as you use a single method on `Queryable`. On my machine, that will add roughly 100 KB of uncompressed IL to a Blazor app.
2. We can annotate each individual `Queryable` method with a `DynamicDependency` attribute to the corresponding `Enumerable` method. That means only the `Enumerable` methods that are being used though `Queryable` will be preserved.
    - There is a bit of a maintenance issue here - we need to remember to add this attribute to any new `Queryable` method, but we can cover that with a unit test.

My plan is to take approach (2) to fix this.

**Comment by stephentoub** [comment]
> My plan is to take approach (2) to fix this.

That sounds like the right one.

For the future, have we explored the idea of allowing the linker to execute code in an assembly?  I'm imagining if this scenario could instead be solved in the future with a variation on (1):
```C#
[DynamicallyAccessedMembersCallback(nameof(GetCorrespondingEnumerableMethod))]
public static class Queryable
{
    ...

    private static MemberInfo GetCorrespondingEnumerableMethod(MemberInfo source) =>
        typeof(Enumerable).GetMethod(source.Name); // something more complicated, but you get the idea
}
```
Just spitballing.

**Comment by stephentoub** [comment]
> but we can cover that with a unit test

FWIW, we already have similar unit tests about correspondance between Queryable and Enumerable:
https://github.com/dotnet/runtime/blob/e87d2470773e48991b8e1d965d49b3ef03b303fe/src/libraries/System.Linq.Queryable/tests/Queryable.cs#L102-L112

**Comment by ElderJames** [comment]
> @ElderJames - do you have a branch of the repo that is targeting `net5.0`?

The `AntDesign.Docs.Wasm` in master branch currently is targeting `net5.0`. 
You can see the exception when it publish to github pages.

https://ant-design-blazor.github.io/en-US/components/table

**Comment by ElderJames** [comment]
> For a 5.0 app, you can turn off trimming by configuring `<PublishTrimmed>false</PublishTrimmed>` in your project file.



> > Thanks @lewing ,
> > I tried, but it was still the exception.
> 
> Whoops, I gave the wrong instructions, try [#42536 (comment)](https://github.com/dotnet/runtime/issues/42536#issuecomment-696802055) instead

It works now when disabled the trimming..

**Comment by eerhardt** [comment]
Re-opening for backport to 5.0.

**Comment by jeffhandley** [comment]
Fixed in 5.0 RC2 with https://github.com/dotnet/runtime/pull/42653.
